File: src/main/java/com/ricedotwho/rsa/module/impl/dungeon/AutoTerms.java
package com.ricedotwho.rsa.module.impl.dungeon;

import com.google.common.collect.Lists;
import com.google.common.primitives.Shorts;
import com.google.common.primitives.SignedBytes;
import com.ricedotwho.rsa.component.impl.TickFreeze;
import com.ricedotwho.rsa.event.impl.RawTickEvent;
import com.ricedotwho.rsa.module.impl.dungeon.terminals.*;
import com.ricedotwho.rsm.RSM;
import com.ricedotwho.rsm.data.Colour;
import com.ricedotwho.rsm.event.api.SubscribeEvent;
import com.ricedotwho.rsm.event.impl.client.InputPollEvent;
import com.ricedotwho.rsm.event.impl.client.PacketEvent;
import com.ricedotwho.rsm.event.impl.game.ClientTickEvent;
import com.ricedotwho.rsm.event.impl.render.Render2DEvent;
import com.ricedotwho.rsm.event.impl.render.Render3DEvent;
import com.ricedotwho.rsm.event.impl.world.WorldEvent;
import com.ricedotwho.rsm.module.Module;
import com.ricedotwho.rsm.module.api.Category;
import com.ricedotwho.rsm.module.api.ModuleInfo;
import com.ricedotwho.rsm.ui.clickgui.settings.impl.*;
import com.ricedotwho.rsm.utils.ChatUtils;
import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
import lombok.Getter;
import net.minecraft.client.KeyMapping;
import net.minecraft.client.Minecraft;
import net.minecraft.client.multiplayer.ClientPacketListener;
import net.minecraft.core.NonNullList;
import net.minecraft.network.HashedStack;
import net.minecraft.network.protocol.game.*;
import net.minecraft.world.entity.player.Input;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.inventory.AbstractContainerMenu;
import net.minecraft.world.inventory.Slot;
import net.minecraft.world.item.ItemStack;
import org.joml.Vector2d;

import java.util.Arrays;
import java.util.List;

@Getter
@ModuleInfo(aliases = "AutoTerms", id = "AutoTerms", category = Category.DUNGEONS)
public class AutoTerms extends Module {
    private long lastClickTime = 0L;
    private boolean clickedWindow = false;
    private boolean firstClick = true;
    private Terminal terminal;

    private AbstractContainerMenu terminalContainer;
    private final ClickedSlotsTracker clickedSlotsTracker;
    private final TerminalRenderer terminalRenderer;

    private TerminalState predictedState = null;

    private final NumberSetting firstClickDelay = new NumberSetting("First Click Delay", 200d, 600d, 400d, 5d);
    private final NumberSetting delay = new NumberSetting("Delay", 100d, 250d, 150d, 5d);
    private final NumberSetting breakThreshold = new NumberSetting("Break Threshold", 200d, 800d, 500d, 10d);

    private final BooleanSetting melodySkip = new BooleanSetting("Melody Skip", true);
    private final BooleanSetting melodySkipFirst = new BooleanSetting("Don't Skip First", true);

    private final GroupSetting invWalkGroup = new GroupSetting("Invwalk");
    private final BooleanSetting doInvwalk = new BooleanSetting("Enabled", false);
    private final ModeSetting style = new ModeSetting("Style", "Items", Arrays.asList("Solver", "Items"));

    private final BooleanSetting renderTitles = new BooleanSetting("Render title thing", true);
    private final BooleanSetting renderClicksLeft = new BooleanSetting("Render clicks left", true);
    private final ColourSetting titleColour = new ColourSetting("Title Colour", new Colour(96,31,158));
    private final ColourSetting remainingColour = new ColourSetting("Remaining Colour", new Colour(96,31,158));
    private final ColourSetting clicksColour = new ColourSetting("Clicks Colour", new Colour(0, 191, 0));
    @Getter private static final ColourSetting solutionColour = new ColourSetting("Solution Colour", new Colour(0, 150, 0));
    @Getter private static final ColourSetting oppositeColour = new ColourSetting("Opposite Colour", new Colour(0, 0, 150));
    @Getter private static final ColourSetting orderColour1 = new ColourSetting("Order Colour 1", new Colour(0, 150, 0));
    @Getter private static final ColourSetting orderColour2 = new ColourSetting("Order Colour 2", new Colour(150, 150, 0));
    @Getter private static final ColourSetting orderColour3 = new ColourSetting("Order Colour 3", new Colour(150, 0, 0));
    private final NumberSetting gap = new NumberSetting("Gap", 0, 3, 1.5, 0.01);
    private final BooleanSetting textShadow = new BooleanSetting("Text Shadow", false);

    private final ModeSetting moveDelayMode = new ModeSetting("Mode Delay", "Stop Inputs", List.of("Stop Inputs", "Freeze"));
    private final NumberSetting melodyMoveDelay = new NumberSetting("Melody Move Delay", 0, 500, 300, 50);

    private final DragSetting termTitle = new DragSetting("Term Title", new Vector2d(10, 10), new Vector2d(150, 15));
    private final DragSetting clicksText = new DragSetting("Clicks Text", new Vector2d(10, 10), new Vector2d(150, 15));
    private final DragSetting gui = new DragSetting("Visualiser Gui", new Vector2d(551, 330), new Vector2d(144, 80));


    private int melodyMoveCounter = 0;


    public AutoTerms() {
        this.clickedSlotsTracker = new ClickedSlotsTracker();
        this.terminalRenderer = new TerminalRenderer();
        registerProperty(
                firstClickDelay,
                delay,
                breakThreshold,
                melodySkip,
                melodySkipFirst,
                invWalkGroup,
                gui,
                termTitle,
                clicksText
        );

        invWalkGroup.add(
                doInvwalk,
                style,
                renderTitles,
                renderClicksLeft,
                titleColour,
                remainingColour,
                clicksColour,
                solutionColour,
                oppositeColour,
                gap,
                textShadow,
                moveDelayMode,
                melodyMoveDelay
        );
    }

    @SubscribeEvent
    public void onLoadWorld(WorldEvent.Load event) {
        close();
    }

    @SubscribeEvent
    public void onRenderGui(Render2DEvent event) {
        try {
           if (!isInTerm() || !this.doInvwalk.getValue()) return;

           float width = 9 * 16f;
           int slots = TerminalRenderer.getGuiSlotCount(this.terminalContainer.getType());
           float height = (float) (Math.floor(slots / 9f) * 16);

           if (this.style.is("Items")) {
               gui.renderScaledGFX(event.getGfx(), () -> terminalRenderer.renderItems(event.getGfx(), this.terminal), width, height);
           } else {
               gui.renderScaled(event.getGfx(), () -> terminalRenderer.renderSolver(this.gap.getValue().floatValue(), this.terminal), width, height);
           }
       } catch (Exception e) {
           throw new RuntimeException(e);
       }
    }

    @SubscribeEvent
    public void render(Render3DEvent.Last event) {
        // issues with race conditions
        if (!isInTerm() || terminal instanceof Melody) return;

        if (terminal.shouldSolve() && !terminal.isSolved()) {
            terminal.solve();
        }

        if (!terminal.isSolved()) return;

        if (predictedState != null) {
            TerminalState newState = this.terminal.getCurrentState();
            if (!predictedState.matches(newState)) {
//                ChatUtils.chat("First click detected!");
//                ChatUtils.chat("Old : " + predictedState.getHash());
//                ChatUtils.chat("new : " + newState.getHash());
                this.firstClick = true;
                this.lastClickTime = System.currentTimeMillis();
                this.clickedSlotsTracker.clear();
            }
            this.predictedState = null;
        }


        if (firstClick && (System.currentTimeMillis() - lastClickTime < firstClickDelay.getValue())) return;

        if (System.currentTimeMillis() - lastClickTime < delay.getValue()) return;

        if (System.currentTimeMillis() - lastClickTime > breakThreshold.getValue()) {
            clickedWindow = false;
        }

        // Why is there another check here?
        if (!isInTerm() || clickedWindow) return;

        if (!terminal.isSolved()) return;

        Solution solution = terminal.getSolution();

        if (solution.getLength() < 1) return;

        sendWindowClick(solution.getNext());
        lastClickTime = System.currentTimeMillis();
        clickedWindow = true;
        firstClick = false;
    }


    // Need to change this for inv walk
    private static void sendWindowClick(int windowID, SolutionClick click, Player player, AbstractContainerMenu abstractContainerMenu) {
        if (windowID != abstractContainerMenu.containerId) {
            ChatUtils.chat("Window ID mismatch!");
            return;
        }

        ClientPacketListener connection = Minecraft.getInstance().getConnection();
        if (connection == null) return;

        NonNullList<Slot> nonNullList = abstractContainerMenu.slots;
        int l = nonNullList.size();
        List<ItemStack> list = Lists.newArrayListWithCapacity(l);

        for (Slot slot : nonNullList) {
            list.add(slot.getItem().copy());
        }

        abstractContainerMenu.clicked(click.index(), click.button(), click.type(), player);
//        ChatUtils.chat("ClickType : " + click.type().name());
//        ChatUtils.chat("Button : " + click.button());
        Int2ObjectMap<HashedStack> int2ObjectMap = new Int2ObjectOpenHashMap<>();

        for (int m = 0; m < l; m++) {
            ItemStack itemStack = list.get(m);
            ItemStack itemStack2 = nonNullList.get(m).getItem();
            if (!ItemStack.matches(itemStack, itemStack2)) {
                int2ObjectMap.put(m, HashedStack.create(itemStack2, connection.decoratedHashOpsGenenerator()));
            }
        }

        HashedStack hashedStack = HashedStack.create(abstractContainerMenu.getCarried(), connection.decoratedHashOpsGenenerator());
        connection.send(new ServerboundContainerClickPacket(windowID, abstractContainerMenu.getStateId(), Shorts.checkedCast(click.index()), SignedBytes.checkedCast(click.button()), click.type(), int2ObjectMap, hashedStack));
    }

    public void sendWindowClick(SolutionClick click) {
        if (Minecraft.getInstance().player == null) return;
        if (!isInTerm() || click.index() < 0 || click.index() >= terminal.getType().getSlotCount()) return;
        // Make some checks
        if (this.terminal instanceof StartsWith || this.terminal instanceof Colors)
            this.clickedSlotsTracker.clickSlot(this.terminalContainer.getSlot(click.index()));
        sendWindowClick(terminal.getWindowID(), click, Minecraft.getInstance().player, this.terminalContainer);
    }

    // This works for strafe but not for forwards and backwards for some reason
    @SubscribeEvent
    public void onPollInput(InputPollEvent event) {
        if (this.melodyMoveCounter < 1) return;

        if (Minecraft.getInstance().screen == null && !this.isInTerm()) {
            this.melodyMoveCounter = 0;
            return;
        }

        Input oldInputs = event.getClientInput();
        Input newInputs = new Input(false, false, false, false, false, oldInputs.shift(), false);
        event.getInputConsumer().accept(newInputs);

        this.melodyMoveCounter--;
    }

    @SubscribeEvent
    public void onTick(ClientTickEvent.Start event) {
        if (!isInTerm()) {
            firstClick = true;
            this.clickedSlotsTracker.clear();
            lastClickTime = System.currentTimeMillis();
        }
    }

    @SubscribeEvent
    public void onRawTick(RawTickEvent event) {
        if (isInTerm() && terminal instanceof Melody melody) {
            if (melody.onTickStart(this)) {
                if (this.moveDelayMode.is("Freeze")) {
                    TickFreeze.freeze(this.melodyMoveDelay.getValue().longValue(), true);
                } else {
                    this.melodyMoveCounter = (this.melodyMoveDelay.getValue().intValue() / 50);
                }
            }
        }
    }

    @SubscribeEvent
    public void onReceivePacket(PacketEvent.Receive event) {
        if (event.getPacket() instanceof ClientboundOpenScreenPacket packet) {
            if (packet.getContainerId() < 1 || packet.getContainerId() > 100) return;
            if (Minecraft.getInstance().player == null) return;

            TerminalState predictionState = new TerminalState(null, 0);
            if (this.terminal != null && this.terminal.isSolved()) {
                predictionState = this.terminal.getNextState();
            }

            this.terminalContainer = packet.getType().create(packet.getContainerId(), Minecraft.getInstance().player.getInventory());

            this.terminal = Terminal.fromPacket(packet, terminalContainer);
            if (this.terminal == null) {
                this.terminalContainer = null;
                return;
            }

            this.predictedState = predictionState;
            this.clickedWindow = false;
            this.terminalRenderer.newWindow(terminalContainer);

            if (doInvwalk.getValue()) event.setCancelled(true);
            return;
        }

        if (isInTerm() && event.getPacket() instanceof ClientboundContainerSetSlotPacket packet) {
            if (packet.getContainerId() == 0 || packet.getContainerId() != this.terminalContainer.containerId) return;
            terminalContainer.setItem(packet.getSlot(), packet.getStateId(), packet.getItem());
            terminal.loadSlot(packet);

            if (doInvwalk.getValue()) event.setCancelled(true);
            return;
        }


        if (isInTerm() && event.getPacket() instanceof ClientboundContainerClosePacket packet) {
            if (packet.getContainerId() != terminalContainer.containerId) {
                ChatUtils.chat("Container ID mismatch on close!");
            }

            this.close();
            if (doInvwalk.getValue()) event.setCancelled(true);
            return;
        }

        if (isInTerm() && event.getPacket() instanceof ClientboundSetCursorItemPacket packet) {
            if (doInvwalk.getValue()) event.setCancelled(true);
            return;
        }

        if (isInTerm() && event.getPacket() instanceof ClientboundContainerSetContentPacket packet) {
            if (packet.containerId() != 0 && doInvwalk.getValue()) event.setCancelled(true);
            return;
        }
    }

    private void close() {
        this.terminal = null;
        this.terminalRenderer.close();
        this.terminalContainer = null;
        this.predictedState = null;
        this.firstClick = true;
        this.lastClickTime = System.currentTimeMillis();
        this.clickedSlotsTracker.clear();
    }

    @SubscribeEvent
    public void onSendPacket(PacketEvent.Send event) {
        if (isInTerm() && event.getPacket() instanceof ServerboundContainerClosePacket packet) {
            this.close();
            return;
        }
    }

    public static boolean isInTerminal() {
        return RSM.getModule(AutoTerms.class).isInTerm();
    }

    private boolean isInTerm() {
        return this.terminal != null && terminalContainer != null;
    }
}

File: src/main/kotlin/rsa/module/impl/dungeon/boss/p3/terminals/auto/AutoTerms.kt
package rsa.module.impl.dungeon.boss.p3.terminals.auto

import com.ricedotwho.rsa.event.impl.RawTickEvent
import com.ricedotwho.rsa.module.impl.dungeon.boss.p3.terminals.auto.terminals.ClickedSlotsTracker
import com.ricedotwho.rsa.module.impl.dungeon.boss.p3.terminals.auto.terminals.Colors
import com.ricedotwho.rsa.module.impl.dungeon.boss.p3.terminals.auto.terminals.Melody
import com.ricedotwho.rsa.module.impl.dungeon.boss.p3.terminals.auto.terminals.SolutionClick
import com.ricedotwho.rsa.module.impl.dungeon.boss.p3.terminals.auto.terminals.StartsWith
import com.ricedotwho.rsa.module.impl.dungeon.boss.p3.terminals.auto.terminals.TerminalState
import com.ricedotwho.rsm.RSM
import com.ricedotwho.rsm.component.impl.location.Island
import com.ricedotwho.rsm.component.impl.location.Location
import com.ricedotwho.rsm.event.api.EventPriority
import com.ricedotwho.rsm.event.api.SubscribeEvent
import com.ricedotwho.rsm.event.impl.client.InputPollEvent
import com.ricedotwho.rsm.event.impl.client.PacketEvent
import com.ricedotwho.rsm.event.impl.game.ClientTickEvent
import com.ricedotwho.rsm.event.impl.render.Render3DEvent
import com.ricedotwho.rsm.event.impl.world.WorldEvent
import com.ricedotwho.rsm.module.Module
import com.ricedotwho.rsm.module.api.Category
import com.ricedotwho.rsm.module.api.ModuleInfo
import com.ricedotwho.rsm.ui.clickgui.settings.group.GroupSetting
import com.ricedotwho.rsm.ui.clickgui.settings.impl.BooleanSetting
import com.ricedotwho.rsm.ui.clickgui.settings.impl.MultiBoolSetting
import com.ricedotwho.rsm.ui.clickgui.settings.impl.NumberSetting
import com.ricedotwho.rsm.utils.DungeonUtils
import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap
import net.minecraft.client.gui.screens.MenuScreens
import net.minecraft.client.gui.screens.inventory.AbstractContainerScreen
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.network.HashedStack
import net.minecraft.network.chat.Component
import net.minecraft.network.protocol.common.ClientboundPingPacket
import net.minecraft.network.protocol.game.ClientboundContainerClosePacket
import net.minecraft.network.protocol.game.ClientboundContainerSetContentPacket
import net.minecraft.network.protocol.game.ClientboundContainerSetDataPacket
import net.minecraft.network.protocol.game.ClientboundContainerSetSlotPacket
import net.minecraft.network.protocol.game.ClientboundHorseScreenOpenPacket
import net.minecraft.network.protocol.game.ClientboundMerchantOffersPacket
import net.minecraft.network.protocol.game.ClientboundOpenScreenPacket
import net.minecraft.network.protocol.game.ClientboundSetCursorItemPacket
import net.minecraft.network.protocol.game.ServerboundContainerClickPacket
import net.minecraft.world.entity.player.Input
import net.minecraft.world.entity.player.Player
import net.minecraft.world.inventory.AbstractContainerMenu
import net.minecraft.world.inventory.MenuType
import net.minecraft.world.item.ItemStack
import rsa.RSA
import rsa.module.getValue
import rsa.module.impl.dungeon.boss.p3.terminals.auto.terminals.Terminal
import rsa.module.provideDelegate

@ModuleInfo(aliases = ["AutoTerms"], id = "AutoTerms", category = Category.DUNGEONS)
object AutoTerms : Module() {
  private var lastClickTime = 0L
  private var clickedWindow = false
  private var firstClick = true
  var terminal: Terminal? = null
    private set

  private var lastPingTicks = 100

  var terminalContainer: AbstractContainerMenu? = null
    private set
  val clickedSlotsTracker = ClickedSlotsTracker()

  private var predictedState: TerminalState? = null

  private val firstClickDelay by NumberSetting("First Click Delay", 200.0, 600.0, 400.0, 5.0)
  private val delay by NumberSetting("Delay", 100.0, 250.0, 150.0, 5.0)
  private val breakThreshold by NumberSetting("Break Threshold", 200.0, 800.0, 500.0, 10.0)

  @JvmStatic
  val terminals: Map<String, Boolean> by MultiBoolSetting(
    "Terminals",
    listOf("Colours", "Melody", "Numbers", "Red Green", "Rubix", "Starts With"),
    listOf("Colours", "Melody", "Numbers", "Red Green", "Rubix", "Starts With")
  )

  val melodySkip: Boolean by BooleanSetting("Melody Skip", true)
  val melodySkipFirst: Boolean by BooleanSetting("Don't Skip First", true)
  val announceMelody: Boolean by BooleanSetting("Announce Melody", true)

  private val noLimbo by BooleanSetting("No Limbo", true)
  private val invWalkGroup by GroupSetting("Invwalk", InvWalk)

  @SubscribeEvent
  private fun onWorldLoad(event: WorldEvent.Load) {
    close()
    lastPingTicks = 100
  }

  @SubscribeEvent
  private fun render(event: Render3DEvent.Last) {
    val terminal = terminal ?: return
    if (!isInTerm() || terminal is Melody) return

    if (terminal.shouldSolve && !terminal.isSolved) {
      terminal.solve()
    }

    if (!terminal.isSolved) return

    predictedState?.let {
      val newState = terminal.getCurrentState()
      if (!it.matches(newState)) {
        firstClick = true
        lastClickTime = System.currentTimeMillis()
        clickedSlotsTracker.clear()
      }
      predictedState = null
    }

    if (!terminal.isEnabled) return

    // todo: DO NOT DO THIS ON RENDER!

    val lastClick = System.currentTimeMillis() - lastClickTime
    val delay = if (firstClick) firstClickDelay else delay
    if (lastClick < delay.toDouble()) return
    if (lastClick > breakThreshold.toDouble()) {
      clickedWindow = true
    }

    // Why is there another check here?
    if (!isInTerm() || clickedWindow) return

    if (!terminal.isSolved) return
    val solution = terminal.solution ?: return
    if (solution.length < 1) return

    sendWindowClick(solution.next)
    lastClickTime = System.currentTimeMillis()
    clickedWindow = true
    firstClick = false
  }

  @SubscribeEvent
  private fun onPollInput(event: InputPollEvent) {
    if (invWalkGroup.melodyMoveCounter < 1) return
    if (mc.screen == null && !isInTerm()) {
      invWalkGroup.melodyMoveCounter = 0
      return
    }

    val oldInputs = event.clientInput
    val newInputs = Input(false, false, false, false, false, oldInputs.shift, false)
    event.input.apply(newInputs)

    --invWalkGroup.melodyMoveCounter
  }

  @SubscribeEvent
  private fun onTick(event: ClientTickEvent.Start) {
    --lastPingTicks
    if (!isInTerm()) {
      firstClick = true
      clickedSlotsTracker.clear()
      lastClickTime = System.currentTimeMillis()
    }
  }

  @SubscribeEvent
  private fun onRawTick(event: RawTickEvent) {
    val terminal = terminal ?: return
    if (!isInTerm() || terminal !is Melody || !terminal.isEnabled) return
    if (terminal.onTickStart(this)) {
      invWalkGroup.onMelodyClick()
    }
  }

  @SubscribeEvent
  private fun onPlayerTick(event: ClientTickEvent.Player) {
    val player = mc.player ?: return
    if (Location.getArea() != Island.Dungeon || !DungeonUtils.isPositionInF7Boss(player.position()) || !isInTerm() || !invWalkGroup.isEnabled) return
    if (lastPingTicks < 0) event.isCancelled = true
  }

  // If a gui is open request is sent at the same time as term aura sends a click packet while not in term,
  // if the original gui opens first, the term gui will open after the client has opened it

  /// This should run before {@link Terminals#onPacket(PacketEvent.Receive)}
  @SubscribeEvent(priority = EventPriority.HIGH)
  private fun onReceivePacket(event: PacketEvent.Receive) {
    val cancel = !invWalkGroup.invwalkMaybeFix
    when (val packet = event.packet) {
      is ClientboundPingPacket -> {
        lastPingTicks = 5
      }

      is ClientboundOpenScreenPacket -> {
        if (packet.containerId !in 1..100) return
        val player = mc.player ?: return

        val predState = terminal?.run { if (isSolved) getNextState() else null } ?: TerminalState(null, 0)

        val terminalContainer = packet.type.create(packet.containerId, player.inventory)
        this.terminalContainer = terminalContainer
        terminal = Terminal.fromPacket(packet, terminalContainer)
        if (terminal == null) {
          this.terminalContainer = null
          return
        }

        // should run after?
        val screen = mc.screen
        if (screen is AbstractContainerScreen<*> && screen.menu.containerId != 0) {
          // o7 Balding
          mc.setScreen(null)
        }

        predictedState = predState
        clickedWindow = false
        invWalkGroup.terminalRenderer.newWindow(terminalContainer)

        if (invWalkGroup.isEnabled) {
          event.isCancelled = true
          if (invWalkGroup.invwalkMaybeFix) {
            setContainerMenu(packet.type, packet.containerId, packet.title)
          }
        }
      }

      is ClientboundContainerSetSlotPacket -> {
        val (terminal, terminalContainer) = (terminal ?: return) to (terminalContainer ?: return)
        if (packet.containerId == 0 || packet.containerId != terminalContainer.containerId) return
        terminalContainer.setItem(packet.slot, packet.stateId, packet.item)
        terminal.loadSlot(packet)
        if (invWalkGroup.isEnabled && cancel) event.isCancelled = true
      }

      is ClientboundContainerClosePacket -> {
        val terminalContainer = terminalContainer ?: return
        if (packet.containerId != terminalContainer.containerId) {
          RSA.chat("Container ID mismatch on close!")
          close()
          return
        }

        close()
        if (invWalkGroup.isEnabled) event.isCancelled = true
      }

      is ClientboundSetCursorItemPacket -> {
        if (!isInTerm()) return
        if (invWalkGroup.isEnabled && cancel) event.isCancelled = true
      }

      is ClientboundContainerSetContentPacket -> {
        if (!isInTerm()) return
        if (packet.containerId != 0 && invWalkGroup.isEnabled && cancel) event.isCancelled = true
      }

      is ClientboundHorseScreenOpenPacket -> {
        if (!isInTerm()) return
        reset()
      }

      is ClientboundContainerSetDataPacket -> {
        if (!isInTerm()) return
        if (packet.containerId != 0 && invWalkGroup.isEnabled && cancel) event.isCancelled = true
      }

      is ClientboundMerchantOffersPacket -> {
        if (!isInTerm()) return
        reset()
      }
    }
  }

  private fun <T : AbstractContainerMenu> setContainerMenu(menuType: MenuType<T>, i: Int, component: Component) {
    val screenConstructor = MenuScreens.getConstructor(menuType) ?: return RSA.logger.warn(
      "Failed to create screen for menu type: {}", BuiltInRegistries.MENU.getKey(menuType)
    )
    val inventory = mc.player?.inventory ?: return
    val screen = screenConstructor.create(menuType.create(i, inventory), inventory, component)
    mc.player?.containerMenu = screen.menu
  }

  private fun sendWindowClick(
    windowId: Int, click: SolutionClick, player: Player, containerMenu: AbstractContainerMenu
  ) {
    if (windowId != containerMenu.containerId) {
      RSA.chat("Window ID mismatch!")
      return
    }

    val conn = mc.connection ?: return
    val nonNullList = containerMenu.slots
    val list = ArrayList<ItemStack>(nonNullList.size)
    nonNullList.forEach { slot ->
      list.add(slot.item.copy())
    }

    containerMenu.clicked(click.index(), click.button(), click.type(), player)

    val int2ObjectMap = Int2ObjectOpenHashMap<HashedStack>()
    nonNullList.forEachIndexed { index, slot ->
      val itemStack = list[index]
      val itemStack2 = slot.item
      if (!ItemStack.matches(itemStack, itemStack2)) {
        int2ObjectMap[index] = HashedStack.create(itemStack2, conn.decoratedHashOpsGenenerator())
      }
    }

    val hashedStack = HashedStack.create(containerMenu.carried, conn.decoratedHashOpsGenenerator())
    conn.send(
      ServerboundContainerClickPacket(
        windowId,
        containerMenu.stateId,
        click.index().toShort(),
        click.button().toByte(),
        click.type(),
        int2ObjectMap,
        hashedStack
      )
    )
  }

  fun sendWindowClick(click: SolutionClick) {
    val terminal = terminal ?: return
    val terminalContainer = terminalContainer ?: return
    val player = mc.player ?: return
    if (click.index() < 0 || click.index() >= terminal.type.slotCount) return
    if (terminal is StartsWith || terminal is Colors) {
      clickedSlotsTracker.clickSlot(terminalContainer.getSlot(click.index()))
    }
    sendWindowClick(terminal.windowId, click, player, terminalContainer)
  }

  private fun close() {
    terminal = null
    invWalkGroup.terminalRenderer.close()
    terminalContainer = null
    predictedState = null
    firstClick = true
    lastClickTime = System.currentTimeMillis()
    clickedSlotsTracker.clear()
  }

  @JvmStatic
  fun isInTerminal() = RSM.getModule(this.javaClass).isInTerm()

  fun isInTerm() = terminal != null && terminalContainer != null
}
